moquette-0.17/broker/src/main/java/io/moquette/broker/SessionEventLoopGroup.java

```diff
@@ -0,0 +1,111 @@
+        }
+        this.sessionExecutors = new SessionEventLoop[eventLoops];
+        for (int i = 0; i < eventLoops; i++) {
+            SessionEventLoop newLoop = new SessionEventLoop(this.sessionQueues[i]);
+            newLoop.setName(sessionLoopName(i));
+            newLoop.setUncaughtExceptionHandler((loopThread, ex) -> {
+                // executed in session loop thread
+                // collect the exception thrown to later re-throw
+                loopThrownExceptions.put(loopThread.getName(), ex);
+
+                // This is done in asynch from another thread in BrokerInterceptor
+                interceptor.notifyLoopException(new InterceptExceptionMessage(ex));
+            });
+            newLoop.start();
+            this.sessionExecutors[i] = newLoop;
+        }
+    }
+
+    int targetQueueOrdinal(String clientId) {
+        return Math.abs(clientId.hashCode()) % this.eventLoops;
+    }
+
+    private String sessionLoopName(int i) {
+        return "Session Executor " + i;
+    }
+
+    String sessionLoopThreadName(String clientId) {
+        final int targetQueueId = targetQueueOrdinal(clientId);
+        return sessionLoopName(targetQueueId);
+    }
+
+    /**
+     * Route the command to the owning SessionEventLoop
+     */
+    public PostOffice.RouteResult routeCommand(String clientId, String actionDescription, Callable<String> action) {
+        SessionCommand cmd = new SessionCommand(clientId, action);
+
+        if (clientId == null) {
+            LOG.warn("Routing collision for action [{}]", actionDescription);
+            return PostOffice.RouteResult.failed(null, "Seems awaiting new route feature completion, skipping.");
+        }
+
+        final int targetQueueId = targetQueueOrdinal(cmd.getSessionId());
+        LOG.debug("Routing cmd [{}] for session [{}] to event processor {}", actionDescription, cmd.getSessionId(), targetQueueId);
+        final FutureTask<String> task = new FutureTask<>(() -> {
+            cmd.execute();
+            cmd.complete();
+            return cmd.getSessionId();
+        });
+        if (Thread.currentThread() == sessionExecutors[targetQueueId]) {
+            SessionEventLoop.executeTask(task);
+            return PostOffice.RouteResult.success(clientId, cmd.completableFuture());
+        }
+        if (this.sessionQueues[targetQueueId].offer(task)) {
+            return PostOffice.RouteResult.success(clientId, cmd.completableFuture());
+        } else {
+            LOG.warn("Session command queue {} is full executing action {}", targetQueueId, actionDescription);
+            return PostOffice.RouteResult.failed(clientId);
+        }
+    }
+
+    public void terminate() {
+        for (SessionEventLoop processor : sessionExecutors) {
+            processor.interrupt();
+        }
+        for (SessionEventLoop processor : sessionExecutors) {
+            try {
+                processor.join(5_000);
+            } catch (InterruptedException ex) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

**Remember**: Do this only for InterruptedException. Doing this for other types of exceptions may give rise to subtle bugs.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)